### ISSUE TITLE
HollowConsumer.Builder cannot be created in Kotlin

### ISSUE THREAD

**Issue by apottere** [Netflix/hollow#252]
Using Kotlin `1.2.16`, I just updated from hollow `2.9.4` to `3.1.0`.  It looks like one of the things changed between those versions was the signature of the `HollowConsumer.Builder` class, now:
```
public class HollowConsumer {
    ...
    public static class Builder<B extends HollowConsumer.Builder> {
        ...
    }
    ...
}
```

In Java, creating a new consumer builder with `new HollowConsumer.Builder()` compiles just fine, but in Kotlin with `HollowConsumer.Builder()` the compiler isn't able to infer argument types.  Given the following test file `Test.kt`:
```
package com.test

import com.netflix.hollow.api.consumer.HollowConsumer

val test = HollowConsumer.Builder()
```

The Kotlin compiler throws:

```
Error:(5, 27) Kotlin: Type inference failed: Not enough information to infer parameter B in constructor Builder<B : HollowConsumer.Builder<(raw) HollowConsumer.Builder<*>>!>()
Please specify it explicitly.
```

After reading [this](https://discuss.kotlinlang.org/t/self-types/371) and [this](https://www.sitepoint.com/self-types-with-javas-generics/), it seems like it's not possible to declare a type explicitly:

> But this exacerbates the problem of convoluted types considerably. Now it’s not even possible to declare a Node because the compiler always expects another type parameter:
> ```
> // doesn't compile
> // the fourth `Node` is not
> // within its type bounds of `Node<Node>`
> Node<Node<Node<Node>>> node = new Node<>();
> ```

This means the consumer cannot be created from Kotlin at all.  In our project, we've worked around the issue by providing a static factory method in Java that we can call from Kotlin:
```
public class Utils {
	static HollowConsumer.Builder newHollowConsumerBuilder() {
		return new HollowConsumer.Builder();
	}
}
```

Could something like this be added to the library for Kotlin users?

**Comment by DanielThomas** [comment]
I wonder if the new type inference algorithm might help with this:

https://youtu.be/MyljSWm0Y_k?t=1306

If something is valid in Java, if Kotlin can't interop with it feels like a Kotlin bug.

**Comment by PaulSandoz** [comment]
Agreed, i would be reluctant to distort the API due to type inference issues be it in Kotlin or say Scala. It may be possible to start from `HollowConsumer.withBlobRetriever`. That works for me in the Kotlin REPL.

**Comment by DanielThomas** [comment]
I found the same - that's how I worked around this.

**Comment by apottere** [comment]
It is just a Kotlin issue.  On the other hand, while recursive generics *do* work in java if you don't specify any type arguments, it's confusing for consumers if they think they have to add type arguments when creating a class (e.g. if their IDE autocompletes empty brackets and expects you to fill them in).  It's also very common for builders in java to be hidden behind a `Type.newBuilder()` method, which is what I looked for before I even started trying to create a builder with `new`.

I don't think it would distort the API very much, if at all, and I'd be willing to create a PR for it.

**Comment by apottere** [comment]
Using `withBlobRetriever` won't work, as the method has the same generic arguments as the class does.  It actually looks like there's a discrepancy between that and `withBlobStore`, where one includes the arguments and one doesn't:
```
    public static <B extends HollowConsumer.Builder<B>> HollowConsumer.Builder<B> withBlobRetriever(HollowConsumer.BlobRetriever blobRetriever) {
        HollowConsumer.Builder builder = new Builder();
        return builder.withBlobRetriever(blobRetriever);
    }

    public static HollowConsumer.Builder withLocalBlobStore(File localBlobStoreDir) {
        HollowConsumer.Builder builder = new Builder();
        return builder.withLocalBlobStore(localBlobStoreDir);
    }
```

I could use `withLocalBlobStore`, but I don't actually want a blob store so I would have to set it to `null` once I had a builder, which seems suboptimal.  It also seems like the type arguments can be removed from `withBlobRetriever`, since AFAICT they aren't necessary.

**Comment by PaulSandoz** [comment]
The signatures of `withBlobRetriever` and `withLocalBlobStore` were corrected in Hollow 3.4.0 (#295)

**Comment by apottere** [comment]
Sorry, I was looking at the source for `3.3.0` at the time.  Thanks for letting me know.

Since you didn't address the other comment, I'm assuming you're not open to adding a `newBuilder()` method?
